### ISSUE TITLE
Empty connection and import ids are unreachable

### ISSUE THREAD

**Issue by kerimoyle** [cellml/libcellml#641]
Now that empty connections are valid (see [here](https://cellml-specification.readthedocs.io/en/latest/reference/formal_and_informative/specB15.html)) there is no way to retrieve the id attribute of an empty connection element:

```
<connection component_1="component1" component_2="component2" id="youCantDoNothingWithMe" /> 
```

The way we retrieve a `connection` id attribute is using the child `map_variables` pairs of variables or through other variable-based functions, so where those variables don't exist, we're stuck?
```
    /**
     * @brief Get the equivalent connection id for this equivalence.
     *
     * Get the connection id set for the equivalence defined with the given variables.
     * The variables are commutative.  If no connection id is set the empty string is returned.
     *
     * If the two variables are not equivalent the empty string is returned.
     *
     * @param variable1 Variable one of the equivalence.
     * @param variable2 Variable two of the equivalence.
     * @return the @c std::string connection id.
     */
    static std::string equivalenceConnectionId(const VariablePtr &variable1, const VariablePtr &variable2);

    /**
     * @brief Get the equivalent connection id for this equivalence.
     *
     * Get the connection id set for the equivalence defined by the this variable and the given one.
     * If no connection id is set the empty string is returned.
     *
     * If the two variables are not equivalent the empty string is returned.
     *
     * @param equivalentVariable The variable this variable is equivalent to.
     * @return The @c std::string id of the equivalence if found otherwise returns the empty string.
     */
    std::string equivalentConnectionId(const VariablePtr &equivalentVariable) const;
```

**Comment by nickerso** [comment]
This is really a more general issue for objects in a CellML model that have no direct representation in the libcellml data model, right? `encapsulation` and `import` elements at least, and maybe more...

I'd currently see this as a very low priority issue to resolve.

**Comment by kerimoyle** [comment]
We can get to the ids of `encapsulation` and `import` items, but we can't get to empty `connection` ones:  
```
model->encapsulationId();
model->component(0)->importSource()->id();
model->units(0)->importSource()->id();
```
It's not so much a matter of representation (as you say, there are plenty of things we don't represent at present), just of access to data that's in the CellML file, but that we can't "read" at the moment.

**Comment by nickerso** [comment]
I doubt we can get to empty `import` or `encapsulation` elements either, right?

**Comment by kerimoyle** [comment]
Encapsulations are ok, this test passes:
```
TEST(Parser, emptyEncapsulationWithId)
{
    const std::string in =
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"model_name\">\n"
        "  <encapsulation id = \"encaps_id\" />\n"
        "</model>\n";

    const std::vector<std::string> expectedIssues = {
        "Encapsulation in model 'model_name' does not contain any child elements.",
    };

    libcellml::ParserPtr p = libcellml::Parser::create();
    auto m = p->parseModel(in);
    EXPECT_EQ_ISSUES(expectedIssues, p);
    EXPECT_EQ(libcellml::Issue::Level::WARNING, p->issue(0)->level());
    EXPECT_EQ("encaps_id", m->encapsulationId());
}
```

**Comment by kerimoyle** [comment]
.. but we can't get to imports either:
```
TEST(Parser, emptyImportWithId)
{
    const std::string in =
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"model_name\">\n"
        "  <import id = \"import_id\" />\n"
        "</model>\n";

    libcellml::ParserPtr p = libcellml::Parser::create();
    auto m = p->parseModel(in);
    EXPECT_EQ("import_id", ??? can't get here either ... );
}
```

**Comment by kerimoyle** [comment]
Closed by #642
